lighthouse-core/lib/minify-trace.js

```diff
@@ -0,0 +1,150 @@
+ */
+'use strict';
+
+/* eslint-disable no-console */
+
+/**
+ * @fileoverview Minifies a trace by removing unnecessary events, throttling screenshots, etc.
+ *  See the following files for necessary events:
+ *    - lighthouse-core/computed/trace-of-tab.js
+ *    - lighthouse-core/computed/page-dependency-graph.js
+ *    - lighthouse-core/lib/dependency-graph/cpu-node.js
+ *    - lighthouse-core/lib/traces/tracing-processor.js
+ */
+
+const TracingProcessor = require('./traces/tracing-processor');
+
+const toplevelTaskNames = new Set([
+  'RunTask', // m71+
+  'ThreadControllerImpl::RunTask', // m69-70
+  'ThreadControllerImpl::DoWork', // m66-68
+  'TaskQueueManager::ProcessTaskFromWorkQueue', // m65 and below
+]);
+
+const traceEventsToAlwaysKeep = new Set([
+  'Screenshot',
+  'TracingStartedInBrowser',
+  'TracingStartedInPage',
+  'navigationStart',
+  'ParseAuthorStyleSheet',
+  'ParseHTML',
+  'PlatformResourceSendRequest',
+  'ResourceSendRequest',
+  'ResourceReceiveResponse',
+  'ResourceFinish',
+  'ResourceReceivedData',
+  'EventDispatch',
+]);
+
+const traceEventsToKeepInToplevelTask = new Set([
+  // Needed for CPU node timing simulations
+  'Layout',
+  // All of these are needed to create graph relationships
+  'TimerInstall',
+  'TimerFire',
+  'InvalidateLayout',
+  'ScheduleStyleRecalculation',
+  'EvaluateScript',
+  'XHRReadyStateChange',
+  'FunctionCall',
+  'v8.compile',
+  'ParseAuthorStyleSheet',
+  'ResourceSendRequest',
+]);
+
+const traceEventsToKeepInProcess = new Set([
+  ...toplevelTaskNames,
+  ...traceEventsToKeepInToplevelTask,
+  'firstPaint',
+  'firstContentfulPaint',
+  'firstMeaningfulPaint',
+  'firstMeaningfulPaintCandidate',
+  'loadEventEnd',
+  'domContentLoadedEventEnd',
+]);
+
+/**
+ * @param {LH.TraceEvent[]} events
+ */
+function filterOutUnnecessaryTasksByNameAndDuration(events) {
+  const {pid} = TracingProcessor.findMainFrameIds(events);
+
+  return events.filter(evt => {
+    // if (toplevelTaskNames.has(evt.name)) return true;
+    // if (toplevelTaskNames.has(evt.name) && evt.dur < 1000) return false;
+    if (evt.pid === pid && traceEventsToKeepInProcess.has(evt.name)) return true;
+    return traceEventsToAlwaysKeep.has(evt.name);
+  });
+}
+
+/**
+ * Filters out tasks that are not within a toplevel task.
+ * @param {LH.TraceEvent[]} events
+ */
+function filterOutOrphanedTasks(events) {
+  const toplevelRanges = events
+    .filter(evt => toplevelTaskNames.has(evt.name))
+    .map(evt => [evt.ts, evt.ts + evt.dur]);
+
+  /** @param {LH.TraceEvent} e */
+  const isInToplevelTask = e => toplevelRanges.some(([start, end]) => e.ts >= start && e.ts <= end);
+
+  return events.filter((evt, index) => {
+    if (!traceEventsToKeepInToplevelTask.has(evt.name)) return true;
+    if (!isInToplevelTask(evt)) return false;
+
+    if (evt.ph === 'B') {
+      const endEvent = events.slice(index).find(e => e.name === evt.name && e.ph === 'E');
```

Comment:
looks slow but I guess is already slow? :)

Reply:
indeed it is, but this isn't exactly mission critical code and the difference isn't massive :)